ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/TraceCall.java

```diff
@@ -0,0 +1,185 @@
+
+    if (maybeBlockHeader.isEmpty()) {
+      return new JsonRpcErrorResponse(requestContext.getRequest().getId(), BLOCK_NOT_FOUND);
+    }
+
+    final Set<TraceTypeParameter.TraceType> traceTypes = traceTypeParameter.getTraceTypes();
+    final DebugOperationTracer tracer = new DebugOperationTracer(buildTraceOptions(traceTypes));
+    final Optional<TransactionSimulatorResult> maybeSimulatorResult =
+        transactionSimulator.process(
+            JsonCallParameterUtil.validateAndGetCallParams(requestContext),
+            buildTransactionValidationParams(),
+            tracer,
+            maybeBlockHeader.get());
+
+    if (maybeSimulatorResult.isEmpty()) {
+      return new JsonRpcErrorResponse(requestContext.getRequest().getId(), INTERNAL_ERROR);
+    }
+
+    final TransactionTrace transactionTrace =
+        new TransactionTrace(
+            maybeSimulatorResult.get().getTransaction(),
+            maybeSimulatorResult.get().getResult(),
+            tracer.getTraceFrames());
+
+    final Block block = blockchainQueries.get().getBlockchain().getChainHeadBlock();
+
+    final TraceCallResult.Builder builder = TraceCallResult.builder();
+
+    transactionTrace
+        .getResult()
+        .getRevertReason()
+        .ifPresentOrElse(
+            revertReason -> builder.output(revertReason.toHexString()),
+            () -> builder.output(maybeSimulatorResult.get().getOutput().toString()));
+
+    if (traceTypes.contains(TraceTypeParameter.TraceType.STATE_DIFF)) {
+      new StateDiffGenerator()
+          .generateStateDiff(transactionTrace)
+          .forEachOrdered(stateDiff -> builder.stateDiff((StateDiffTrace) stateDiff));
+    }
+
+    if (traceTypes.contains(TraceTypeParameter.TraceType.TRACE)) {
+      FlatTraceGenerator.generateFromTransactionTrace(
+              protocolSchedule, transactionTrace, block, new AtomicInteger(), false)
+          .forEachOrdered(trace -> builder.addTrace((FlatTrace) trace));
+    }
+
+    if (traceTypes.contains(VM_TRACE)) {
+      new VmTraceGenerator(transactionTrace)
+          .generateTraceStream()
+          .forEachOrdered(vmTrace -> builder.vmTrace((VmTrace) vmTrace));
+    }
+
+    return new JsonRpcSuccessResponse(
+        requestContext.getRequest().getId(),
+        MAPPER_IGNORE_REVERT_REASON.valueToTree(builder.build()));
+  }
+
+  private TransactionValidationParams buildTransactionValidationParams() {
+    return ImmutableTransactionValidationParams.builder()
+        .from(TransactionValidationParams.transactionSimulator())
+        .build();
+  }
+
+  private TraceOptions buildTraceOptions(final Set<TraceTypeParameter.TraceType> traceTypes) {
+    return new TraceOptions(
+        traceTypes.contains(TraceTypeParameter.TraceType.STATE_DIFF),
+        false,
+        traceTypes.contains(TraceTypeParameter.TraceType.TRACE)
+            || traceTypes.contains(TraceTypeParameter.TraceType.VM_TRACE));
+  }
+
+  private Optional<BlockHeader> resolveBlockHeader(
```

Comment:
Not sure it's doable but why we are not extending the `AbstractBlockParameterMethod` in order to have access to `findResultByParamType`method ? When we extend this class we have all we need to have the block from the request (ex: TraceReplayBlockTransactions)

Reply:
@daniellehrner

Reply:
I did it and I could indeed remove resolveBlockHeader because of it